tsdb/block.go

```diff
@@ -443,7 +447,15 @@ func (r blockIndexReader) LabelValues(name string, matchers ...*labels.Matcher)
 		return st, errors.Wrapf(err, "block: %s", r.b.Meta().ULID)
 	}
 
-	return labelValuesWithMatchers(r, name, matchers...)
+	return labelValuesWithMatchers(r.ir, name, matchers...)
+}
+
+func (r blockIndexReader) LabelNames(matchers ...*labels.Matcher) ([]string, error) {
+	if len(matchers) == 0 {
+		return r.b.LabelNames()
```

Comment:
This calls `Block.LabelNames` which calls the `index.Reader.LabelNames` and it's the only call there and it has no matchers. You'll find a TODO later to implement matchers there, but actually we should just move the matchers evaluation from here to `index.Reader.LabelNames`.

I would love to do that by moving `labelNamesWithMatchers` to `index.LabelNamesWithMatchers`, however, that's a bigger refactor since we'd need to move the `PostingsForMatchers` and even the `IndexReader` interface there too. I also think that it makes sense, but it's just out of the scope of this PR.